**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC4029 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: <font color="#0679c8">Rodrigo González Rodríguez</font>
*   MATRÍCULA: <font color="#0679c8">A01796932</font>


---

En esta actividad usarás la base de datos relacional `classicmodels` (MySQL), compuesta por las siguientes tablas:

*   `Customers`: almacena los datos de los clientes.
*   `Products`: almacena una lista de modelos de coches a escala.
*   `ProductLines`: almacena una lista de categorías de líneas de productos.
*   `Orders`: almacena los pedidos de venta realizados por los clientes.
*   `OrderDetails`: almacena elementos de línea de pedidos de ventas para cada pedido de ventas.
*   `Payments`: almacena los pagos realizados por los clientes en función de sus cuentas.
*   `Employees`: almacena toda la información de los empleados, así como la estructura de la organización, como quién informa a quién.
*   `Offices`: almacena los datos de la oficina de ventas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.


Recuerda que:


*   Una **clave primaria** es un atributo (o conjunto) que identifica unívocamente a cada registro en la tabla.
*   Una **clave foránea** (externa o ajena) es un atributos (o conjunto) en una tabla que es una clave primaria en otra (o posiblemente la misma) tabla.
*   Las **relaciones** son las líneas que conectan una tabla con otra y el extremo determina la cardinalidad. Las relaciones con línea continua (identificadora) representan una transformación donde la clave primaria de una tabla pasa a ser foránea y primaria (al mismo tiempo) de otra. Las relaciones con línea discontinua (no identificadora) representan una transformación donde la clave primaria de una tabla pasa a ser sólo foránea en otra.

# **Parte 1**. SQLAlchemy y SQL básico

In [1]:
pip install pymysql

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.0/45.0 kB 2.2 MB/s eta 0:00:00


In [2]:
import sqlalchemy as sqla
import pymysql
import pandas as pd

1.	Crea el motor `sqlalchemy`, con el método `create_engine()` y una conexión con `connect()` como se muestra a continuación:

In [3]:
# Crear el motor (dialecto://usuarioBD:clave@ipHostDBMS:puerto/esquema
db = sqla.create_engine('mysql+pymysql://mnaTC4029User:mnaTC4029Pass!@20.106.217.214:3306/classicmodels', pool_recycle=3600)

# Crea una conexión para luego invocar declaraciones SQL
conn = db.connect()

Escribe las consultas en SQL para obtener:

2.	La información de las líneas de productos.

<font color="#0679c8">Tenemos 110 productos, dentro de sus atributos tenemos: codigo, nombre, linea de producto, escala, vendedor, descripcion, cantidad en inventario, precio de compra y precio sugerido de venta</font>

In [4]:
query = "SELECT * FROM products;"
products = pd.read_sql(query, conn)
products

productCode                            productName   productLine  \
0      S10_1678  1969 Harley Davidson Ultimate Chopper   Motorcycles   
1      S10_1949               1952 Alpine Renault 1300  Classic Cars   
2      S10_2016                  1996 Moto Guzzi 1100i   Motorcycles   
3      S10_4698   2003 Harley-Davidson Eagle Drag Bike   Motorcycles   
4      S10_4757                    1972 Alfa Romeo GTA  Classic Cars   
..          ...                                    ...           ...   
105   S700_3505                            The Titanic         Ships   
106   S700_3962                         The Queen Mary         Ships   
107   S700_4002              American Airlines: MD-11S        Planes   
108    S72_1253                       Boeing X-32A JSF        Planes   
109    S72_3212                             Pont Yacht         Ships   

    productScale              productVendor  \
0           1:10            Min Lin Diecast   
1           1:10    Classic Metal Creations   
2           1:10   Highway 66 Mini Classics   
3           1:10          Red Start Diecast   
4           1:10    Motor City Art Classics   
..           ...                        ...   
105        1:700   Carousel DieCast Legends   
106        1:700  Welly Diecast Productions   
107        1:700        Second Gear Diecast   
108         1:72    Motor City Art Classics   
109         1:72       Unimax Art Galleries   

                                    productDescription  quantityInStock  \
0    This replica features working kickstand, front...             7933   
1    Turnable front wheels; steering function; deta...             7305   
2    Official Moto Guzzi logos and insignias, saddl...             6625   
3    Model features, official Harley Davidson logos...             5582   
4    Features include: Turnable front wheels; steer...             3252   
..                                                 ...              ...   
105  Completed model measures 19 1/2 inches long, 9...             1956   
106  Exact replica. Wood and Metal. Many extras inc...             5088   
107  Polished finish. Exact replia with official lo...             8820   
108  10" Wingspan with retractable landing gears.Co...             4857   
109  Measures 38 inches Long x 33 3/4 inches High. ...              414   

     buyPrice    MSRP  
0       48.81   95.70  
1       98.58  214.30  
2       68.99  118.94  
3       91.02  193.66  
4       85.68  136.00  
..        ...     ...  
105     51.09  100.17  
106     53.63   99.31  
107     36.27   74.03  
108     32.77   49.66  
109     33.30   54.60  

[110 rows x 9 columns]

3.	La información de los empleados ordenados por nombre (`firstName`).

<font color="#0679c8"> Tenemos 23 empleados, dentro de sus atributos podemos encontrar: Numero de empleado, nombre y apellido, numero de entension, correo, el codigo de la oficina a la que pertenece, el codigo del empleado a quien le reporta, y su puesto.</font>

In [5]:
query = "SELECT * FROM employees ORDER BY firstName;"
employees = pd.read_sql(query, conn)
employees

employeeNumber   lastName firstName extension  \
0             1611     Fixter      Andy      x101   
1             1143        Bow   Anthony     x5428   
2             1504      Jones     Barry      x102   
3             1002     Murphy     Diane     x5800   
4             1286      Tseng  Foon Yue     x2248   
5             1323     Vanauf    George     x4102   
6             1102     Bondur    Gerard     x5408   
7             1370  Hernandez    Gerard     x2028   
8             1076   Firrelli      Jeff     x9273   
9             1188   Firrelli     Julie     x2173   
10            1501       Bott     Larry     x2311   
11            1165   Jennings    Leslie     x3291   
12            1166   Thompson    Leslie     x4065   
13            1337     Bondur      Loui     x6493   
14            1621      Nishi      Mami      x101   
15            1702     Gerard    Martin     x2312   
16            1056  Patterson      Mary     x4611   
17            1401   Castillo    Pamela     x2759   
18            1612      Marsh     Peter      x102   
19            1216  Patterson     Steve     x4334   
20            1619       King       Tom      x103   
21            1088  Patterson   William     x4871   
22            1625       Kato   Yoshimi      x102   

                              email officeCode  reportsTo  \
0      afixter@classicmodelcars.com          6     1088.0   
1         abow@classicmodelcars.com          1     1056.0   
2       bjones@classicmodelcars.com          7     1102.0   
3      dmurphy@classicmodelcars.com          1        NaN   
4       ftseng@classicmodelcars.com          3     1143.0   
5      gvanauf@classicmodelcars.com          3     1143.0   
6      gbondur@classicmodelcars.com          4     1056.0   
7    ghernande@classicmodelcars.com          4     1102.0   
8    jfirrelli@classicmodelcars.com          1     1002.0   
9    jfirrelli@classicmodelcars.com          2     1143.0   
10       lbott@classicmodelcars.com          7     1102.0   
11   ljennings@classicmodelcars.com          1     1143.0   
12   lthompson@classicmodelcars.com          1     1143.0   
13     lbondur@classicmodelcars.com          4     1102.0   
14      mnishi@classicmodelcars.com          5     1056.0   
15     mgerard@classicmodelcars.com          4     1102.0   
16   mpatterso@classicmodelcars.com          1     1002.0   
17   pcastillo@classicmodelcars.com          4     1102.0   
18      pmarsh@classicmodelcars.com          6     1088.0   
19  spatterson@classicmodelcars.com          2     1143.0   
20       tking@classicmodelcars.com          6     1088.0   
21  wpatterson@classicmodelcars.com          6     1056.0   
22       ykato@classicmodelcars.com          5     1621.0   

                jobTitle  
0              Sales Rep  
1     Sales Manager (NA)  
2              Sales Rep  
3              President  
4              Sales Rep  
5              Sales Rep  
6    Sale Manager (EMEA)  
7              Sales Rep  
8           VP Marketing  
9              Sales Rep  
10             Sales Rep  
11             Sales Rep  
12             Sales Rep  
13             Sales Rep  
14             Sales Rep  
15             Sales Rep  
16              VP Sales  
17             Sales Rep  
18             Sales Rep  
19             Sales Rep  
20             Sales Rep  
21  Sales Manager (APAC)  
22             Sales Rep

4.	Los países donde hay oficinas (sin duplicar).

<font color="#0679c8">USA, Francia, Japon, Australia y UK</font>

In [6]:
query = "SELECT DISTINCT(country) FROM offices;"
offices = pd.read_sql(query, conn)
offices

country
0        USA
1     France
2      Japan
3  Australia
4         UK

5.	El nombre y teléfono de los clientes de la ciudad de Nueva York (*NYC*).

<font color="#0679c8">Tenemos 5 contactos en NYC </font>

In [7]:
query = "SELECT contactFirstName,contactLastName,phone FROM customers WHERE city = 'NYC';"
customers = pd.read_sql(query, conn)
customers

contactFirstName contactLastName       phone
0             Kwai             Lee  2125557818
1             Jeff           Young  2125557413
2          Michael           Frick  2125551500
3            Maria       Hernandez  2125558493
4               Yu            Choi  2125551957

6.	El código y nombre de los productos del vendedor *Gearbox Collectibles* que tengan menos de 1000 unidades en stock.

<font color="#0679c8">Solo hay dos productos : P-51-D Mustang (S18_2581) y 1928 Mercedes-Benz SSK (S18_2795) </font>

In [8]:
query = "SELECT productCode, productName FROM products WHERE productVendor = 'Gearbox Collectibles' AND quantityInStock < 1000;"
products_2 = pd.read_sql(query, conn)
products_2

productCode             productName
0    S18_2581          P-51-D Mustang
1    S18_2795  1928 Mercedes-Benz SSK

7.	Los tres productos más caros, desde el punto de visto de los comercializadores (`buyPrice`).

<font color="#0679c8">Los tres productos mas caros son : 1962 LanciaA Delta 16V, 1998 Chrysler plymouth Prowler y 1952 Alpine Renault 1300, con precios entre 98.58 y 103.42 </font>

In [9]:
query = "SELECT productCode, productName,buyPrice FROM products ORDER BY buyPrice DESC LIMIT 3;"
products_3 = pd.read_sql(query, conn)
products_3

productCode                     productName  buyPrice
0    S10_4962          1962 LanciaA Delta 16V    103.42
1    S18_2238  1998 Chrysler Plymouth Prowler    101.51
2    S10_1949        1952 Alpine Renault 1300     98.58

8.	La cantidad de productos por línea de producto (no las existencias en inventario)

<font color="#0679c8">Tenemos 7 lineas de productos, siendo "Autos Clasicos" la mayor con 38 productos distintos, y "Trenes" la menor, con solo 3 productos </font>

In [10]:
query = "SELECT productLine, count(*) AS 'Quantity' FROM products GROUP BY productLine;"
products_4 = pd.read_sql(query, conn)
products_4

productLine  Quantity
0      Classic Cars        38
1       Motorcycles        13
2            Planes        12
3             Ships         9
4            Trains         3
5  Trucks and Buses        11
6      Vintage Cars        24

9.	La cantidad de empleados por país (tomando en cuenta la ubicación de la oficina).

<font color="#0679c8">Tenemos 23 empleados en total, el 43% (10) estan localizados en USA, 22% (5) en Francia, 17% (4) en Austalia y el resto esta distribuidor en Japon y UK equitativamente.</font>


---




In [11]:
query = "SELECT country, count(*) AS 'employees' FROM employees JOIN offices USING (officeCode) GROUP BY country;"
employees_2 = pd.read_sql(query, conn)
employees_2

country  employees
0  Australia          4
1     France          5
2      Japan          2
3         UK          2
4        USA         10

10.	El promedio de los pagos de cada uno de los clientes de España (sin incluir aquellos que no poseen ningún pago).

<font color="#0679c8">En España tenemos 5 clientes que realizaron pagos, donde el promedio de sus pagos esta entre 22,840 y 55,056</font>


---



In [12]:
query = "SELECT customerName,AVG(amount) AS 'average_payment' FROM customers INNER JOIN payments USING (customerNumber) WHERE country = 'Spain' GROUP BY customerName;"
customers_2 = pd.read_sql(query, conn)
customers_2

customerName  average_payment
0                 CAF Imports     23375.570000
1  Corrida Auto Replicas, Ltd     37480.030000
2          Enaco Distributors     22840.156667
3      Euro+ Shopping Channel     55056.844615
4  Iberia Gift Imports, Corp.     25493.925000

# **Parte 2**. Manipulación de datos con Pandas

11.	Carga las tablas empleadas en dataframes con el mismo nombre y resuelve las consultas anteriores con las funciones de Pandas (NO con SQL). Cuida no sobreescribir los dataframes originales al resolver las consultas. Debes obtener los mismos resultados que con SQL.

<font color="#0679c8">Cargamos las tablas usadas en las consultas como dataframes</font>


In [13]:
products_df = pd.read_sql("SELECT * FROM products;", conn)
employees_df = pd.read_sql("SELECT * FROM employees;", conn)
offices_df = pd.read_sql("SELECT * FROM offices;", conn)
customers_df = pd.read_sql("SELECT * FROM customers;", conn)
payments_df = pd.read_sql("SELECT * FROM payments;", conn)

<font color="#0679c8">2.	La información de las líneas de productos.
</font>


In [14]:
# SELECT * FROM products;
products_df

productCode                            productName   productLine  \
0      S10_1678  1969 Harley Davidson Ultimate Chopper   Motorcycles   
1      S10_1949               1952 Alpine Renault 1300  Classic Cars   
2      S10_2016                  1996 Moto Guzzi 1100i   Motorcycles   
3      S10_4698   2003 Harley-Davidson Eagle Drag Bike   Motorcycles   
4      S10_4757                    1972 Alfa Romeo GTA  Classic Cars   
..          ...                                    ...           ...   
105   S700_3505                            The Titanic         Ships   
106   S700_3962                         The Queen Mary         Ships   
107   S700_4002              American Airlines: MD-11S        Planes   
108    S72_1253                       Boeing X-32A JSF        Planes   
109    S72_3212                             Pont Yacht         Ships   

    productScale              productVendor  \
0           1:10            Min Lin Diecast   
1           1:10    Classic Metal Creations   
2           1:10   Highway 66 Mini Classics   
3           1:10          Red Start Diecast   
4           1:10    Motor City Art Classics   
..           ...                        ...   
105        1:700   Carousel DieCast Legends   
106        1:700  Welly Diecast Productions   
107        1:700        Second Gear Diecast   
108         1:72    Motor City Art Classics   
109         1:72       Unimax Art Galleries   

                                    productDescription  quantityInStock  \
0    This replica features working kickstand, front...             7933   
1    Turnable front wheels; steering function; deta...             7305   
2    Official Moto Guzzi logos and insignias, saddl...             6625   
3    Model features, official Harley Davidson logos...             5582   
4    Features include: Turnable front wheels; steer...             3252   
..                                                 ...              ...   
105  Completed model measures 19 1/2 inches long, 9...             1956   
106  Exact replica. Wood and Metal. Many extras inc...             5088   
107  Polished finish. Exact replia with official lo...             8820   
108  10" Wingspan with retractable landing gears.Co...             4857   
109  Measures 38 inches Long x 33 3/4 inches High. ...              414   

     buyPrice    MSRP  
0       48.81   95.70  
1       98.58  214.30  
2       68.99  118.94  
3       91.02  193.66  
4       85.68  136.00  
..        ...     ...  
105     51.09  100.17  
106     53.63   99.31  
107     36.27   74.03  
108     32.77   49.66  
109     33.30   54.60  

[110 rows x 9 columns]

<font color="#0679c8">3.	La información de los empleados ordenados por nombre (`firstName`).
</font>

In [15]:
# SELECT * FROM employees ORDER BY firstName DESC;
employees_df.sort_values(by='firstName')

employeeNumber   lastName firstName extension  \
17            1611     Fixter      Andy      x101   
5             1143        Bow   Anthony     x5428   
16            1504      Jones     Barry      x102   
0             1002     Murphy     Diane     x5800   
10            1286      Tseng  Foon Yue     x2248   
11            1323     Vanauf    George     x4102   
13            1370  Hernandez    Gerard     x2028   
4             1102     Bondur    Gerard     x5408   
2             1076   Firrelli      Jeff     x9273   
8             1188   Firrelli     Julie     x2173   
15            1501       Bott     Larry     x2311   
6             1165   Jennings    Leslie     x3291   
7             1166   Thompson    Leslie     x4065   
12            1337     Bondur      Loui     x6493   
20            1621      Nishi      Mami      x101   
22            1702     Gerard    Martin     x2312   
1             1056  Patterson      Mary     x4611   
14            1401   Castillo    Pamela     x2759   
18            1612      Marsh     Peter      x102   
9             1216  Patterson     Steve     x4334   
19            1619       King       Tom      x103   
3             1088  Patterson   William     x4871   
21            1625       Kato   Yoshimi      x102   

                              email officeCode  reportsTo  \
17     afixter@classicmodelcars.com          6     1088.0   
5         abow@classicmodelcars.com          1     1056.0   
16      bjones@classicmodelcars.com          7     1102.0   
0      dmurphy@classicmodelcars.com          1        NaN   
10      ftseng@classicmodelcars.com          3     1143.0   
11     gvanauf@classicmodelcars.com          3     1143.0   
13   ghernande@classicmodelcars.com          4     1102.0   
4      gbondur@classicmodelcars.com          4     1056.0   
2    jfirrelli@classicmodelcars.com          1     1002.0   
8    jfirrelli@classicmodelcars.com          2     1143.0   
15       lbott@classicmodelcars.com          7     1102.0   
6    ljennings@classicmodelcars.com          1     1143.0   
7    lthompson@classicmodelcars.com          1     1143.0   
12     lbondur@classicmodelcars.com          4     1102.0   
20      mnishi@classicmodelcars.com          5     1056.0   
22     mgerard@classicmodelcars.com          4     1102.0   
1    mpatterso@classicmodelcars.com          1     1002.0   
14   pcastillo@classicmodelcars.com          4     1102.0   
18      pmarsh@classicmodelcars.com          6     1088.0   
9   spatterson@classicmodelcars.com          2     1143.0   
19       tking@classicmodelcars.com          6     1088.0   
3   wpatterson@classicmodelcars.com          6     1056.0   
21       ykato@classicmodelcars.com          5     1621.0   

                jobTitle  
17             Sales Rep  
5     Sales Manager (NA)  
16             Sales Rep  
0              President  
10             Sales Rep  
11             Sales Rep  
13             Sales Rep  
4    Sale Manager (EMEA)  
2           VP Marketing  
8              Sales Rep  
15             Sales Rep  
6              Sales Rep  
7              Sales Rep  
12             Sales Rep  
20             Sales Rep  
22             Sales Rep  
1               VP Sales  
14             Sales Rep  
18             Sales Rep  
9              Sales Rep  
19             Sales Rep  
3   Sales Manager (APAC)  
21             Sales Rep

<font color="#0679c8">4.	Los países donde hay oficinas (sin duplicar).
</font>


In [16]:
# SELECT DISTINCT(country) FROM offices;
offices_df[['country']].drop_duplicates()

country
0        USA
3     France
4      Japan
5  Australia
6         UK

<font color="#0679c8">5.	El nombre y teléfono de los clientes de la ciudad de Nueva York (*NYC*).
</font>

In [17]:
# SELECT contactFirstName,contactLastName,phone FROM customers WHERE city = 'NYC';
customers_df[customers_df['city'] == 'NYC'][['contactFirstName', 'contactLastName', 'phone']]

contactFirstName contactLastName       phone
9               Kwai             Lee  2125557818
15              Jeff           Young  2125557413
27           Michael           Frick  2125551500
98             Maria       Hernandez  2125558493
105               Yu            Choi  2125551957

<font color="#0679c8">6.	El código y nombre de los productos del vendedor *Gearbox Collectibles* que tengan menos de 1000 unidades en stock.</font>

In [18]:
# SELECT productCode, productName FROM products WHERE productVendor = 'Gearbox Collectibles' AND quantityInStock < 1000;
products_df[(products_df['productVendor'] == 'Gearbox Collectibles') & (products_df['quantityInStock'] < 1000)][['productCode', 'productName']]

productCode             productName
30    S18_2581          P-51-D Mustang
32    S18_2795  1928 Mercedes-Benz SSK

<font color="#0679c8">7.	Los tres productos más caros, desde el punto de visto de los comercializadores (`buyPrice`).</font>

In [19]:
# SELECT productCode, productName, buyPrice FROM products ORDER BY buyPrice DESC LIMIT 3;
products_df.sort_values(by='buyPrice', ascending=False).head(3)[['productCode', 'productName', 'buyPrice']]

productCode                     productName  buyPrice
5     S10_4962          1962 LanciaA Delta 16V    103.42
25    S18_2238  1998 Chrysler Plymouth Prowler    101.51
1     S10_1949        1952 Alpine Renault 1300     98.58

<font color="#0679c8">8.	La cantidad de productos por línea de producto (no las existencias en inventario)</font>

In [20]:
# SELECT productLine, count(*) AS 'Quantity' FROM products GROUP BY productLine;
products_df.groupby('productLine').size()

productLine
Classic Cars        38
Motorcycles         13
Planes              12
Ships                9
Trains               3
Trucks and Buses    11
Vintage Cars        24
dtype: int64

<font color="#0679c8">9.	La cantidad de empleados por país (tomando en cuenta la ubicación de la oficina).</font>


In [21]:
# SELECT country, count(*) AS 'employees' FROM employees JOIN offices USING (officeCode) GROUP BY country;
employees_df.merge(offices_df, on='officeCode').groupby('country').size()

country
Australia     4
France        5
Japan         2
UK            2
USA          10
dtype: int64

<font color="#0679c8">10.	El promedio de los pagos de cada uno de los clientes de España (sin incluir aquellos que no poseen ningún pago).</font>


In [22]:
# SELECT customerName,AVG(amount) AS 'average_payment' FROM customers JOIN payments USING (customerNumber) WHERE country = 'Spain' GROUP BY customerName;
customer_payments_df = customers_df.merge(payments_df, on='customerNumber')[['customerName', 'amount', 'country']]
customer_payments_df[customer_payments_df['country'] == 'Spain'].groupby('customerName').mean('amount')

amount
customerName                            
CAF Imports                 23375.570000
Corrida Auto Replicas, Ltd  37480.030000
Enaco Distributors          22840.156667
Euro+ Shopping Channel      55056.844615
Iberia Gift Imports, Corp.  25493.925000

# **Parte 3**. Cliente de Python Firestore

En esta fase te conectarás a una base de datos no relacional de Firestore desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [23]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

El archivo `veterinary.json` almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios de Firebase. A través de la función `Certificate()`, se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación. Después de eso, tu conexión a Firestore utilizará las reglas de seguridad establecidas para la base de datos y el usuario autenticado.

In [24]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [25]:
import os
DIR = "/content/drive/MyDrive/Colab Notebooks/01 - Ciencia y analítica de datos/3.2 - Acceso BD y manipulacion de datos"
os.chdir(DIR)

In [26]:
cred = credentials.Certificate('veterinary.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

12.	Investiga cómo leer la colección `PET_OWNER` y mostrar su contenido en un dataframe. Asegúrate de incluir el id en el resultado

<font color="#0679c8">Tenemos 10 dueños de mascotas, todos con nombre, apellido, id, y la mayoria con correo y telefono</font>


In [27]:
# El metodo .collection() en firestore se usa para acceder a una base de datos
# El metodo .stream() se usa para poder accesar a cada dato de la coleccion
pet_owner_collection = db.collection('PET_OWNER').stream()
# Crearemos una lista donde el ciclo insertara cada registro
# Convertimos cada registro a un diccionario y lo descomprimimos usando **
# Le agregamos el id de cada registro que no estaba en el formato original de la tabla
data = []
for pet_owner in pet_owner_collection:
    data.append({**pet_owner.to_dict(), 'id': pet_owner.id})
# Al final solo convertimos el diccionario resultante a dataframe
pd.DataFrame(data)

email         phone ownerFirstName ownerLastName  \
0                         None  555-454-3465            Sam        Taylor   
1    miles.trent@somewhere.com          None          Miles         Trent   
2      liz.frier@somewhere.com  555-537-6543            Liz         Frier   
3                         None  555-454-1243          Jenny      Mayberry   
4  'marcha.downs@somewhere.com  555-537-8765         Marsha         Downs   
5                         None  555-454-2354            Ken       Roberts   
6   nigel.melnik@somewhere.com  555-232-5678          Nigel        Melnik   
7  richard.james@somewhere.com  555-537-7654        Richard         James   
8     jim.rogers@somewhere.com  555-232-3456            Jim        Rogers   
9    mary.keenan@somewhere.com  555-232-4567           Mary        Keenan   

                     id  
0  0D6yFSs2eu4nYwf2dnQ0  
1  98357ufgjmWUxjnAuzbz  
2  AFtZincSZxjC4Mcxf9Pf  
3  GYDixDwHjMyihjL8TmsM  
4  IpxTuB6FILhwQFcspLws  
5  QoXBfS1JdRp6BqgLzaAz  
6  eUtZ44lPRbSHXX04EDp7  
7  n2GG9rlfGOmE0P4TcU0T  
8  pDlJlul8EJXO8FBJvFkN  
9  tAlJ74xQ7tT4TUHwZgmH

In [28]:
firebase_admin.delete_app(firebase_admin.get_app())